File: axo_vem/axo_vem/infrastructure/transport/api/controllers/buckets.py
from __future__ import annotations

from typing import Callable

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from pymongo.collection import Collection
from xolo.client.models import UserDTO

from axo_shared import wire
from axo_shared.client import DEFAULT_CHUNK_BYTES
from axo_shared.protocol import Command

from axo_vem.domain.data.repository import BucketOwnerRepository, BucketRepository, DataItemRepository
from axo_vem.infrastructure.transport.zmq_command.endpoint_client import resolve_rpc_uri, send_command


class _BucketCreateRequest(BaseModel):
    name: str
    quota_bytes: int


def _endpoint_rpc_uri(endpoints: Collection, endpoint_id: str) -> str:
    doc = endpoints.find_one({"_id": endpoint_id})
    if doc is None:
        raise HTTPException(status_code=404, detail="endpoint not found")
    router_bind = doc.get("router_bind")
    if not router_bind:
        raise HTTPException(status_code=409, detail="endpoint has no known router_bind")
    return resolve_rpc_uri(router_bind, endpoint_id)


def _leader_rpc_uri(endpoints: Collection, consensus: Collection, fallback: str) -> str:
    """DATA_CHUNK_PUT is never leader-proxied (unlike DATA_REGISTER), so it
    must land on the actual current leader. The DATA_REGISTER response's own
    leader_rpc_uri is unusable here -- it's the leader's own bind string
    (e.g. tcp://0.0.0.0:5555), not a connectable address, and every node
    shares that same string. Resolve the real leader via the consensus
    collection instead (see controllers/consensus.py's identical
    "highest term = current leader" query), falling back to the
    originally-targeted endpoint if no consensus view has been reported yet
    (single-node dev)."""
    doc = consensus.find_one(sort=[("_id", -1)])
    if doc is None or not doc.get("leader_ids"):
        return fallback
    leader_id = doc["leader_ids"][0]
    leader_doc = endpoints.find_one({"_id": leader_id})
    if leader_doc is None or not leader_doc.get("router_bind"):
        return fallback
    return resolve_rpc_uri(leader_doc["router_bind"], leader_id)


def build_router(
    bucket_repository: BucketRepository,
    data_item_repository: DataItemRepository,
    bucket_owner_repository: BucketOwnerRepository,
    endpoints: Collection,
    consensus: Collection,
    current_user_dependency: Callable[..., UserDTO],
    command_timeout_seconds: float = 3.0,
) -> APIRouter:
    router = APIRouter(tags=["data buckets"])

    def _used_bytes(bucket_name: str) -> int:
        return sum(item.total_size for item in data_item_repository.list_by_bucket(bucket_name))

    @router.get("/buckets")
    def list_buckets(
        mine_only: bool = False,
        current_user: UserDTO = Depends(current_user_dependency),
    ):
        """Unfiltered by default -- buckets created before ownership existed
        have no owner row and would otherwise vanish from every listing.
        mine_only=true scopes to buckets this user actually created."""
        buckets = bucket_repository.list()
        if mine_only:
            owned = set(bucket_owner_repository.list_owned(current_user.key))
            buckets = [b for b in buckets if b.name in owned]
        return [
            {
                **bucket.to_dict(),
                "used_bytes": _used_bytes(bucket.name),
                "owner_user_id": bucket_owner_repository.get_owner(bucket.name),
            }
            for bucket in buckets
        ]

    @router.get("/buckets/{name}")
    def get_bucket(name: str, _current_user: UserDTO = Depends(current_user_dependency)):
        bucket = bucket_repository.get(name)
        if bucket is None:
            raise HTTPException(status_code=404, detail="bucket not found")
        items = data_item_repository.list_by_bucket(name)
        return {
            **bucket.to_dict(),
            "used_bytes": sum(item.total_size for item in items),
            "owner_user_id": bucket_owner_repository.get_owner(name),
            "items": [item.to_dict() for item in items],
        }

    @router.post("/endpoints/{endpoint_id}/buckets", tags=["endpoint management"])
    def create_bucket(
        endpoint_id: str,
        body: _BucketCreateRequest,
        current_user: UserDTO = Depends(current_user_dependency),
    ):
        """Proxies BUCKET_REGISTER straight to the target endpoint's ROUTER
        (leader-proxied by the node itself if needed), mirroring
        register_function/assign_virtual_environment's relay shape. Ownership
        is recorded here, directly, synchronously -- not via the node round
        trip (see BucketOwnerRepository)."""
        rpc_uri = _endpoint_rpc_uri(endpoints, endpoint_id)
        command = Command(
            operation=wire.BUCKET_REGISTER,
            content_type="application/json",
            envelope={"name": body.name, "quota_bytes": body.quota_bytes},
        )
        result = send_command(rpc_uri, command, command_timeout_seconds)
        if result.is_err:
            raise HTTPException(status_code=504, detail=str(result.unwrap_err()))
        command_result = result.unwrap()
        if not command_result.ok:
            raise HTTPException(status_code=409, detail=command_result.error)
        bucket_owner_repository.set_owner(body.name, current_user.key)
        return {"endpoint_id": endpoint_id, **(command_result.metadata or {})}

    @router.post("/endpoints/{endpoint_id}/buckets/{bucket}/data", tags=["endpoint management"])
    async def upload_data(
        endpoint_id: str,
        bucket: str,
        key: str = Form(...),
        version: int = Form(...),
        format: str = Form("raw"),
        kind: str = Form("fs"),
        file: UploadFile = File(...),
        _current_user: UserDTO = Depends(current_user_dependency),
    ):
        """Accepts a browser file upload and proxies it into the cluster:
        one DATA_REGISTER relay (leader-proxied by the node), then a loop of
        DATA_CHUNK_PUT relays targeting the resolved leader directly. The
        whole file is read into memory here -- acceptable for the expected
        data sizes; revisit if that becomes a bottleneck."""
        register_rpc_uri = _endpoint_rpc_uri(endpoints, endpoint_id)
        name = f"{bucket}/{key}"
        source = await file.read()
        total_size = len(source)

        register_command = Command(
            operation=wire.DATA_REGISTER,
            content_type="application/json",
            envelope={
                "name": name, "version": version, "format": format, "kind": kind,
                "total_size": total_size, "chunk_bytes": DEFAULT_CHUNK_BYTES,
            },
        )
        result = send_command(register_rpc_uri, register_command, command_timeout_seconds)
        if result.is_err:
            raise HTTPException(status_code=504, detail=str(result.unwrap_err()))
        command_result = result.unwrap()
        if not command_result.ok:
            raise HTTPException(status_code=409, detail=command_result.error)

        total_chunks = command_result.metadata.get("total_chunks", 0)
        chunk_rpc_uri = _leader_rpc_uri(endpoints, consensus, fallback=register_rpc_uri)
        for i in range(total_chunks):
            start = i * DEFAULT_CHUNK_BYTES
            chunk_command = Command(
                operation=wire.DATA_CHUNK_PUT,
                content_type="application/octet-stream",
                envelope={"name": name, "version": version, "chunk_index": i},
                payload=source[start:start + DEFAULT_CHUNK_BYTES],
            )
            chunk_result = send_command(chunk_rpc_uri, chunk_command, command_timeout_seconds)
            if chunk_result.is_err:
                raise HTTPException(status_code=504, detail=str(chunk_result.unwrap_err()))
            chunk_command_result = chunk_result.unwrap()
            if not chunk_command_result.ok:
                raise HTTPException(status_code=409, detail=chunk_command_result.error)

        return {
            "endpoint_id": endpoint_id, "bucket": bucket, "name": name, "version": version,
            "total_size": total_size, "total_chunks": total_chunks,
        }

    @router.get(
        "/endpoints/{endpoint_id}/buckets/{bucket}/data/{key}/{version}/download", tags=["endpoint management"],
    )
    def download_data(
        endpoint_id: str,
        bucket: str,
        key: str,
        version: int,
        _current_user: UserDTO = Depends(current_user_dependency),
    ):
        """Streams a previously-uploaded dataset's raw bytes back to the
        browser -- the read counterpart of upload_data. Reads from the
        resolved leader (same _leader_rpc_uri target upload_data's chunk
        puts use), since that's the one node BlobReplicator's completeness
        gate guarantees is fully present once the item's status is "ready";
        picking a specific complete replica via DATA_INFO's peer state is a
        possible future optimization, not needed today. Raw bytes only, no
        format-aware decoding -- mirrors the dataio design's "the endpoint
        only ever moves raw bytes" principle end to end.

        Validation (item exists, upload complete, endpoint/leader
        resolvable) all happens before StreamingResponse is constructed, so
        those failures still get a clean 404/409/504 -- once streaming
        actually starts, headers are already committed and a failed chunk
        fetch can only abort the connection, not swap in a different status
        code."""
        name = f"{bucket}/{key}"
        item = data_item_repository.get(name, version)
        if item is None:
            raise HTTPException(status_code=404, detail="data not found")
        if item.status != "ready":
            raise HTTPException(status_code=409, detail="data upload is not yet complete")

        register_rpc_uri = _endpoint_rpc_uri(endpoints, endpoint_id)
        chunk_rpc_uri = _leader_rpc_uri(endpoints, consensus, fallback=register_rpc_uri)

        def _iter_chunks():
            for i in range(item.total_chunks):
                chunk_command = Command(
                    operation=wire.DATA_CHUNK_GET,
                    content_type="application/json",
                    envelope={"name": name, "version": version, "chunk_index": i},
                )
                chunk_result = send_command(chunk_rpc_uri, chunk_command, command_timeout_seconds)
                if chunk_result.is_err:
                    raise RuntimeError(f"failed to fetch chunk {i} of {name}:{version}: {chunk_result.unwrap_err()}")
                command_result = chunk_result.unwrap()
                if not command_result.ok:
                    raise RuntimeError(f"failed to fetch chunk {i} of {name}:{version}: {command_result.error}")
                yield command_result.payload

        filename = key.rsplit("/", 1)[-1]
        return StreamingResponse(
            _iter_chunks(),
            media_type="application/octet-stream",
            headers={"Content-Disposition": f'attachment; filename="{filename}"'},
        )

    @router.delete("/endpoints/{endpoint_id}/buckets/{bucket}/data/{key}/{version}", tags=["endpoint management"])
    def delete_data(
        endpoint_id: str,
        bucket: str,
        key: str,
        version: int,
        _current_user: UserDTO = Depends(current_user_dependency),
    ):
        """Proxies DATA_DELETE straight to the target endpoint's ROUTER --
        leader-proxied by the node itself (like DATA_REGISTER/BUCKET_REGISTER
        above), unlike DATA_CHUNK_PUT which needs the special leader
        resolution in _leader_rpc_uri. The deletion itself then replicates
        to every other cluster member via the node's own consensus path."""
        rpc_uri = _endpoint_rpc_uri(endpoints, endpoint_id)
        command = Command(
            operation=wire.DATA_DELETE,
            content_type="application/json",
            envelope={"name": f"{bucket}/{key}", "version": version},
        )
        result = send_command(rpc_uri, command, command_timeout_seconds)
        if result.is_err:
            raise HTTPException(status_code=504, detail=str(result.unwrap_err()))
        command_result = result.unwrap()
        if not command_result.ok:
            raise HTTPException(status_code=409, detail=command_result.error)
        return {"endpoint_id": endpoint_id, "bucket": bucket, "name": f"{bucket}/{key}", "version": version}

    return router

File: axo-ui/app/stores/buckets.ts
import type { BucketCreateRequest, BucketStatusMessage, BucketWithItems, DataBucket } from '~/types/bucket'

export const useBucketsStore = defineStore('buckets', () => {
  const { apiFetch, wsUrl } = useApi()

  const items = ref<DataBucket[]>([])
  const current = ref<BucketWithItems | null>(null)
  const loading = ref(false)
  const error = ref<string | null>(null)
  let socket: WebSocket | null = null

  async function fetchAll() {
    loading.value = true
    error.value = null
    try {
      items.value = await apiFetch<DataBucket[]>('/buckets')
    } catch (err: any) {
      error.value = err.message
    } finally {
      loading.value = false
    }
  }

  async function fetchOne(name: string) {
    loading.value = true
    error.value = null
    try {
      current.value = await apiFetch<BucketWithItems>(`/buckets/${name}`)
    } catch (err: any) {
      error.value = err.message
    } finally {
      loading.value = false
    }
  }

  async function create(endpointId: string, payload: BucketCreateRequest) {
    loading.value = true
    error.value = null
    try {
      await apiFetch(`/endpoints/${endpointId}/buckets`, { method: 'POST', body: payload })
      await fetchAll()
    } catch (err: any) {
      error.value = err.message
      loading.value = false
      throw err
    }
  }

  // `file` is sent as multipart/form-data, mirroring stores/functions.ts::register's
  // code upload -- the browser reads the file into memory and proxies it
  // through axo_vem, which relays DATA_REGISTER + chunked
  // DATA_CHUNK_PUT calls to the cluster on the caller's behalf. Deliberately
  // does NOT refetch the bucket afterward -- that used to race the async
  // node-event -> Kurrent -> projector -> Mongo pipeline the read model
  // depends on (the bug this whole pending/ready + WS design fixes). The
  // new row appears via the "buckets" WS push instead, see connectSocket().
  async function uploadData(
    endpointId: string,
    bucket: string,
    key: string,
    version: number,
    file: File,
    format = 'raw',
    kind = 'fs',
  ) {
    loading.value = true
    error.value = null
    try {
      const formData = new FormData()
      formData.append('key', key)
      formData.append('version', String(version))
      formData.append('format', format)
      formData.append('kind', kind)
      formData.append('file', file)

      await apiFetch(`/endpoints/${endpointId}/buckets/${bucket}/data`, { method: 'POST', body: formData })
    } catch (err: any) {
      error.value = err.message
      throw err
    } finally {
      loading.value = false
    }
  }

  async function deleteData(endpointId: string, bucket: string, key: string, version: number) {
    loading.value = true
    error.value = null
    try {
      await apiFetch(`/endpoints/${endpointId}/buckets/${bucket}/data/${key}/${version}`, { method: 'DELETE' })
      // No local splice here either -- the row disappears via the same WS
      // push's status: "deleted" message, once the deletion has actually
      // replicated, not merely been accepted by one node.
    } catch (err: any) {
      error.value = err.message
      throw err
    } finally {
      loading.value = false
    }
  }

  // Downloads still go through apiFetch's auth headers (unlike the WS
  // connection) since GET /endpoints/.../download is a normal authenticated
  // REST route, just one whose response body is a raw byte stream instead
  // of JSON -- $fetch's default responseType would try to parse it as JSON,
  // so this asks for a Blob instead and drives the browser's own
  // save-file-as flow via a throwaway <a>.
  async function downloadData(endpointId: string, bucket: string, key: string, version: number) {
    error.value = null
    try {
      const blob = await apiFetch<Blob>(
        `/endpoints/${endpointId}/buckets/${bucket}/data/${key}/${version}/download`,
        { responseType: 'blob' },
      )
      const url = URL.createObjectURL(blob)
      const link = document.createElement('a')
      link.href = url
      link.download = key.split('/').pop() || key
      link.click()
      URL.revokeObjectURL(url)
    } catch (err: any) {
      error.value = err.message
      throw err
    }
  }

  function _upsertItem(message: BucketStatusMessage) {
    if (!current.value || message.bucket !== current.value.name) return

    if (message.status === 'deleted') {
      current.value.items = current.value.items.filter(
        item => !(item.name === message.name && item.version === message.version),
      )
      return
    }

    const existing = current.value.items.find(
      item => item.name === message.name && item.version === message.version,
    )
    if (existing) {
      existing.status = message.status
      return
    }
    // Only a "pending" message (the first one for any given item) carries
    // the full shape -- "ready" can only ever be flipping a row that
    // already exists, since pending always lands first. See
    // bucket_handler.py's broadcast payload for why.
    if (message.status === 'pending' && message.format !== undefined) {
      current.value.items.push({
        name: message.name, version: message.version, format: message.format,
        kind: message.kind!, total_size: message.total_size!, total_chunks: message.total_chunks!,
        status: 'pending',
      })
    }
  }

  function connectSocket() {
    if (socket) return
    socket = new WebSocket(wsUrl('/ws/buckets'))
    socket.addEventListener('message', (event) => {
      _upsertItem(JSON.parse(event.data) as BucketStatusMessage)
    })
  }

  function disconnectSocket() {
    socket?.close()
    socket = null
  }

  return {
    items, current, loading, error,
    fetchAll, fetchOne, create, uploadData, deleteData, downloadData,
    connectSocket, disconnectSocket,
  }
})
